src/foolscap/connections/tor.py

```diff
@@ -0,0 +1,153 @@
+        # py2 automatic unicode promotion blows up
+        host = host.encode("ascii")
+        ep = txtorcon.TorClientEndpoint(host, port,
+                                        socks_hostname=self._socks_hostname,
+                                        socks_port=self._socks_port)
+        returnValue( (ep, host) )
+
+
+# note: TorClientEndpoint imports 'reactor' itself, doesn't provide override
+
+class _SocksTor:
+    def __init__(self):
+        _Common.__init__(self)
+        self._connnected = True # no need to call _connect()
+        self._socks_hostname = None
+        self._socks_port = None # use defaults: 9050, then 9150
+
+def default_socks_handler():
+    return _SocksTor()
+
+
+class _LaunchedTor:
+    def __init__(self, reactor, tor_binary=None):
+        _Common.__init__(self)
+        self._reactor = reactor
+        self._tor_binary = tor_binary
+
+    @inlineCallbacks
+    def _connect(self):
+        # create a new Tor
+        config = self.config = txtorcon.TorConfig()
+        if 0:
+            # The default is for launch_tor to create a tempdir itself,
+            # and delete it when done. We only need to set a
+            # DataDirectory if we want it to be persistent.
+            import tempfile
+            datadir = tempfile.mkdtemp()
+            config.DataDirectory = datadir
+
+        #config.ControlPort = allocate_tcp_port() # defaults to 9052
+        #print("setting config.ControlPort to", config.ControlPort)
+        config.SocksPort = allocate_tcp_port()
+        self._socks_hostname = "127.0.0.1"
+        self._socks_port = config.SocksPort
+        print("setting config.SocksPort to", config.SocksPort)
+
+        tpp = yield txtorcon.launch_tor(config, self._reactor,
+                                        tor_binary=self._tor_binary,
+                                        )
+        # gives a TorProcessProtocol with .tor_protocol
+        self._tor_protocol = tpp.tor_protocol
+        print("tp:", self._tor_protocol)
+        returnValue(True)
+
+def launch_tor(reactor, tor_binary=None):
+    """Return a handler which launches a new Tor process (once).
+    - tor_binary: the path to the Tor executable we should use. If None,
+      search $PATH
+    """
+    return _LaunchedTor(reactor, tor_binary)
+
+
+@implementer(IConnectionHintHandler)
+class _ConnectedTor:
+    def __init__(self, reactor, tor_control_endpoint):
+        _Common.__init__(self)
+        self._reactor = reactor
+        assert IStreamClientEndpoint.providedBy(tor_control_endpoint)
+        self._tor_control_endpoint = tor_control_endpoint
+
+    @inlineCallbacks
+    def _connect(self):
+        tor_protocol = yield txtorcon.build_tor_connection(control_endpoint,
+                                                           build_state=False)
+        config = TorConfig(tor_protocol)
+        self._socks_hostname = "127.0.0.1"
+        self._socks_port = config.SocksPort
+
```

Comment:
SocksPort is actually a `StringList` in Tor, I believe. So probably should be `config.SocksPort[0]`. Also, sadly, everything in Tor that's a "list" has no sub-type .. so although these are integer ports, they're all strings from Tor...but should be safe to do `int(config.SocksPort[0])` if you want integer ports. Also there can be zero SocksPorts configured :/

Another approach (as I'm doing with 1.x/controller.Tor in txtorcon) is to give an optional SocksPort to this thing -- so then you: configure it in this tor if you preferred port isn't there already; or choose a random one of the configured ports (e.g. the 0th one); or fail.